scripts/gb_model/chp_utils.py

```diff
@@ -0,0 +1,306 @@
+
+    Parameters
+    ----------
+    n : pypsa.Network
+        The PyPSA network object. Generators must have a 'set' attribute
+        for CHP identification (values: 'CHP', 'PP', 'Store').
+    powerplants : pd.DataFrame
+        Powerplants dataframe including CHP identification.
+    heat_demand_path : str
+        Path to hourly heat demand NetCDF file.
+    heat_to_power_ratio : float
+        Heat-to-power ratio for CHPs (c_b coefficient).
+    min_operation_level : float
+        Minimum operation level (0-1) when CHP is running.
+    shutdown_threshold : float, optional
+        Heat demand threshold below which CHPs can shut down.
+        Default is 0.1 (10% of peak demand).
+
+    Returns
+    -------
+    None
+        Modifies the network object in-place.
+
+    Notes
+    -----
+    This is a simplified CHP model that:
+    - Models CHPs as generators (not Links)
+    - Uses p_min_pu to enforce minimum operation following heat demand
+    - Does not model heat buses or sector coupling
+    - Suitable for electricity-only optimization with CHP constraints
+    - Requires generators to have 'set' attribute for precise CHP identification
+
+    Raises
+    ------
+    ValueError
+        If generators do not have 'set' attribute
+    FileNotFoundError
+        If heat demand file is not found
+    KeyError
+        If required configuration keys are missing
+    """
+    # Identify CHP powerplants
+    chp_plants = identify_chp_powerplants(powerplants)
+
+    if chp_plants.empty:
+        logger.info("No CHP powerplants to constrain.")
+        return
+
+    # Find corresponding generators in the network
+    # CHPs are identified using the 'set' attribute
+    if "set" not in n.generators.columns:
+        logger.error(
+            "'set' attribute not found in generators. "
+            "This is required for CHP identification. "
+            "Ensure attach_conventional_generators() properly sets this attribute."
+        )
+        raise ValueError("'set' attribute missing from generators dataframe")
+
+    chp_generators = n.generators[n.generators["set"] == "CHP"]
+
+    if chp_generators.empty:
+        logger.info(
+            "No CHP generators found in the network. "
+            f"Total generators: {len(n.generators)}, "
+            f"generators by set: {n.generators.groupby('set').size().to_dict()}"
+        )
+        return
+
+    logger.info(
+        f"Applying CHP constraints to {len(chp_generators)} generators "
+        f"with total capacity {chp_generators.p_nom.sum():.1f} MW"
+    )
+
+    # Calculate minimum operation profile
+    # Get buses from matched CHP generators
+    buses = pd.Index(chp_generators["bus"].unique())
+
+    p_min_pu = calculate_chp_minimum_operation(
+        heat_demand_path=heat_demand_path,
+        buses=buses,
+        heat_to_power_ratio=heat_to_power_ratio,
+        min_operation_level=min_operation_level,
+        shutdown_threshold=shutdown_threshold,
+    )
```

Comment:
I think this (which is the lion's share of this util module) should be a script attached to its own rule, generating a timeseries per bus that is then loaded in during `compose_network`. You can use the `fes_powerplants.py` file to get the bus names that have CHPs in them. Then you do the renaming and attaching to the network in `compose_network`. Avoids us having to run this script every time we compose a network. The timeseries data should already be ready.